pkg/stub/handler.go

```diff
@@ -116,61 +116,67 @@ type Handler struct {
+func (h *Handler) StoreCurrentValidConfig(condition v1alpha1.SamplesResourceConditionType, srcfg *v1alpha1.SamplesResource) error {
 
-	if condition == v1alpha1.SamplesExist {
-		cm, err := h.configmapclientwrapper.Get(h.namespace, v1alpha1.SamplesResourceName)
-		if err != nil && !kerrors.IsNotFound(err) {
-			// just return error to sdk for retry
-			return err
-		}
-		if kerrors.IsNotFound(err) {
-			// 4.0 testing showed that we were getting empty ConfigMaps
-			cm = nil
-		}
-		if cm != nil {
-			// just return ... we only update the config map once
-			return nil
-		}
-		cm = &corev1.ConfigMap{}
-		cm.Name = v1alpha1.SamplesResourceName
-		cm.Data = map[string]string{}
-		if len(srcfg.Spec.InstallType) == 0 {
-			cm.Data[installtypekey] = string(v1alpha1.CentosSamplesDistribution)
-		} else {
-			cm.Data[installtypekey] = string(srcfg.Spec.InstallType)
-		}
-		if len(srcfg.Spec.Architectures) == 0 {
-			cm.Data[v1alpha1.X86Architecture] = v1alpha1.X86Architecture
-		} else {
-			for _, arch := range srcfg.Spec.Architectures {
-				switch arch {
-				case v1alpha1.X86Architecture:
-					cm.Data[v1alpha1.X86Architecture] = v1alpha1.X86Architecture
-				case v1alpha1.PPCArchitecture:
-					cm.Data[v1alpha1.PPCArchitecture] = v1alpha1.PPCArchitecture
-				}
+	if condition != v1alpha1.SamplesExist {
+		return nil
+	}
+
+	cm, err := h.configmapclientwrapper.Get(h.namespace, v1alpha1.SamplesResourceName)
+	if err != nil && !kerrors.IsNotFound(err) {
+		// just return error to sdk for retry
+		return err
+	}
+	if kerrors.IsNotFound(err) {
+		// 4.0 testing showed that we were getting empty ConfigMaps
+		cm = nil
+	}
+	if cm != nil {
+		// just return ... we only update the config map once
+		return nil
+	}
+	cm = &corev1.ConfigMap{}
+	cm.Name = v1alpha1.SamplesResourceName
+	cm.Data = map[string]string{}
+	if len(srcfg.Spec.InstallType) == 0 {
+		cm.Data[installtypekey] = string(v1alpha1.CentosSamplesDistribution)
+	} else {
+		cm.Data[installtypekey] = string(srcfg.Spec.InstallType)
+	}
+	if len(srcfg.Spec.Architectures) == 0 {
+		cm.Data[v1alpha1.X86Architecture] = v1alpha1.X86Architecture
+	} else {
+		for _, arch := range srcfg.Spec.Architectures {
+			switch arch {
+			case v1alpha1.X86Architecture:
+				cm.Data[v1alpha1.X86Architecture] = v1alpha1.X86Architecture
+			case v1alpha1.PPCArchitecture:
+				cm.Data[v1alpha1.PPCArchitecture] = v1alpha1.PPCArchitecture
 			}
 		}
-		_, err = h.configmapclientwrapper.Create(h.namespace, cm)
-		if err != nil {
-			// just return error to sdk for retry
-			return err
-		}
 	}
-	return nil
+	_, err = h.configmapclientwrapper.Create(h.namespace, cm)
+	return err
 }
 
 func (h *Handler) SpecValidation(srcfg *v1alpha1.SamplesResource) error {
+	// TODO - the first thing this should do is check that all the config values
+	// are "valid" (the architecture name is known, the distribution name is known, etc)
+	// if that fails, we should immediately error out and set ConfigValid to false.
```

Comment:
> if they specify empty strings for arch/distro, assuming we'll do the default, error out?

no, emptystring is fine.  but invalid values (ie a distribution or arch we don't understand) is not fine.

> at least for distro, the field is InstallType SamplesDistributionType ... so do you mean if they programmatically did SamplesDistributionType("foobar") ?

yes but of course for them it's not programmatic, it's just yaml editing of the resource (oc edit) so the types don't help them.  it's just a string and to them any value is "valid".  Nothing is going to block any valid string from getting set and you processing (or attempting to process) it.

> the arch is a string ... so sure, either add a new SamplesArchitectureType, or do explicit checks so that is it only the two valid

defining a type doesn't matter, for one thing that new "Type" is still just a string.  It's not like it's an enumeration of the valid values.  For another those types don't come into play when the user is editing yaml.  They're a programmatic convenience for your operator code and that's about it.
 
> so essentially just move those check here so the invalid config gets specified ... correct?

i believe so.  Essentially everytime we see a SamplesResource event, the first thing (well one of the first things) we should be doing is confirming the config it specifies is valid.  Valid in terms of standing on its own, and valid in terms of being compatible with the current applied config.

Reply:
OK I will do that in my next follow up PR